src/cargo/ops/cargo_output_metadata.rs

```diff
@@ -0,0 +1,129 @@
+use std::collections::HashMap;
+use std::io::Write;
+use std::fs;
+use std::path::{Path, PathBuf};
+use rustc_serialize::{json, Decoder, Decodable};
+use toml;
+use util::CargoResult;
+use util::config::Config;
+use term::color::BLACK;
+use core::{Source, Package};
+use core::resolver::Resolve;
+use sources::PathSource;
+use ops;
+
+#[derive(RustcDecodable)]
+pub enum OutputFormat {
+    Toml,
+    Json,
+}
+
+/// Where the dependencies should be written to.
+pub enum OutputTo {
+    Path(PathBuf),
+    StdOut,
+}
+
+impl Decodable for OutputTo {
+    fn decode<D: Decoder>(d: &mut D) -> Result<OutputTo, D::Error> {
+        d.read_option(|d, b| {
+            if b {
+                Ok(OutputTo::Path(try!(Decodable::decode(d))))
+            } else {
+                Ok(OutputTo::StdOut)
+            }
+        })
+    }
+}
+
+pub struct OutputMetadataOptions<'a> {
+    pub features: Vec<String>,
+    pub output_format: OutputFormat,
+    pub output_to: OutputTo,
+    pub manifest_path: &'a Path,
+    pub no_default_features: bool,
+    pub target: Option<String>,
+}
+
+/// Loads the manifest, resolves the dependencies of the project to the concrete
+/// used versions - considering overrides - and writes all dependencies in a TOML
+/// format to stdout or the specified file.
+///
+/// The TOML format is e.g.:
+/// ```toml
+/// root = "libA"
+///
+/// [packages.libA]
+/// dependencies = ["libB"]
+/// path = "/home/user/.cargo/registry/src/github.com-1ecc6299db9ec823/libA-0.1"
+/// version = "0.1"
+///
+/// [packages.libB]
+/// dependencies = []
+/// path = "/home/user/.cargo/registry/src/github.com-1ecc6299db9ec823/libB-0.4"
+/// version = "0.4"
+///
+/// [packages.libB.features]
+/// featureA = ["featureB"]
+/// featureB = []
+/// ```
+pub fn output_metadata(opt: OutputMetadataOptions, config: &Config) -> CargoResult<()> {
+    let (resolved_deps, packages) =
+        try!(resolve_dependencies(
+            opt.manifest_path, config, opt.features, opt.target, opt.no_default_features));
+
+    #[derive(RustcEncodable)]
+    struct ExportInfo<'a> {
+        root: String,
```

Comment:
Unfortunately, RustcEncodable impl for Package does not export features information.